testScripts/ObjectDownloadMedia.js

```diff
@@ -0,0 +1,146 @@
+    const out = this.args.out || "./out";
+
+    if(objectId === undefined){
+      if(versionHash===undefined){
+        throw Error("require object-id or object-hash to be provided");
+      }
+      const res = utils.DecodeVersionHash(versionHash);
+      objectId = res.objectId;
+    }
+
+    const dirPath = path.resolve(out);
+    if (!fs.existsSync(dirPath)) {
+      fs.mkdirSync(dirPath, { recursive: true });
+      console.log(`Directory created at ${dirPath}`);
+    } else {
+      throw Error(`Directory already exists at ${dirPath}`);
+    }
+
+
+    const sourcesJsonPath = "offerings." + offeringType + ".media_struct.streams.video.sources[*]";
+    const libraryId = await client.ContentObjectLibraryId({objectId, versionHash});
+    const totalDurationJsonPath = "offerings." + offeringType + ".media_struct.streams.video.duration";
+
+    // get object metadata
+    let metadata = await client.ContentObjectMetadata({
+      libraryId,
+      objectId
+    });
+
+    let sourcesMetadata = JSONPath({
+      json: metadata,
+      path: sourcesJsonPath,
+      wrap: false
+    });
+    if(sourcesMetadata === undefined) {
+      throw new Error("no matching offerings metadata found");
+    }
+
+    let totalDurationMetadata = JSONPath({
+      json: metadata,
+      path: totalDurationJsonPath,
+      wrap: false
+    });
+    if(totalDurationMetadata === undefined) {
+      throw new Error("no matching total duration metadata found");
+    }
+
+    const totalDurationTimeBase = new Fraction(totalDurationMetadata.time_base);
+    const totalDuration = new Fraction(totalDurationMetadata.ts).mul(totalDurationTimeBase);
+    if(startTime < 0 || endTime > totalDuration) {
+      throw new Error(`start or end time provided are not within range: ${totalDuration}s`);
+    }
+
+    const sourcesTimeInfo = sourcesMetadata.map((item) => {
+      // start = item.timeline_start.ts * item.timeline_start.time_base
+      const startTimeBase = new Fraction(item.timeline_start.time_base);
+      const start = new Fraction(item.timeline_start.ts).mul(startTimeBase);
+      // end = item.timeline_start.ts * item.timeline_end.time_base
+      const endTimeBase = new Fraction(item.timeline_end.time_base);
+      const end = new Fraction(item.timeline_end.ts).mul(endTimeBase);
+
+      return {
+        start: start.valueOf(),
+        end: end.valueOf(),
+        source: item.source
+      };
+    });
+
+    let parts = [];
+    sourcesTimeInfo.forEach(item => {
+      if(startTime < item.end && endTime > item.start){
+        parts.push(item.source);
+      }
+    });
+
+    console.log("PARTS:");
+    console.log(parts);
+    console.log();
+
+    console.log("Downloading parts...");
+    for(const partHash of parts) {
+      console.log(partHash);
+      const arrayBuffer = await client.DownloadPart({
+        libraryId,
+        objectId,
+        partHash
+      });
+      let filePath = path.join(dirPath, partHash);
```

Comment:
Might want to prepend an index so that e.g. instead of creating files:
```
hqpeKzJ5bBqKr38dEdNsX16epCwnqmwm5acnRcL13PvQkZhCnR9si
hqpe2NC2ojcbkPLrEbEKfFFSqroeST8Y2wgrJWRWtP61Lg7TfZ7nRT
```

the files wind up being
```
0001.hqpeKzJ5bBqKr38dEdNsX16epCwnqmwm5acnRcL13PvQkZhCnR9si
0002.hqpe2NC2ojcbkPLrEbEKfFFSqroeST8Y2wgrJWRWtP61Lg7TfZ7nRT
```
so that they are sorted in directory listings and user does not have to save part list from output to preserve ordering info.

Alternately you might be able to open a single file handle and write each part to same file handle in append mode, resulting in just one concatenated file being output

Reply:
prepended index to file name.